compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -974,30 +1340,69 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
         }
 
         ty::Dynamic(predicates, _region, kind) => {
-            ty = Ty::new_dynamic(
-                tcx,
-                transform_predicates(tcx, predicates),
-                tcx.lifetimes.re_erased,
-                *kind,
-            );
-        }
-
-        ty::Alias(..) => {
-            ty = transform_ty(
-                tcx,
-                tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), ty),
-                options,
-            );
+            if is_fn_trait(tcx, predicates) {
+                // Transform Fn trait objects into function pointers
+                let fn_sig = ty::Binder::bind_with_vars(
+                    fn_trait_sig(tcx, predicates).unwrap(),
+                    List::empty(),
+                );
+                let fn_ptr = Ty::new_fn_ptr(tcx, fn_sig);
+                // Transform fn_sig inputs and output
+                ty = transform_ty(tcx, fn_ptr, options);
+            } else {
+                ty = Ty::new_dynamic(
+                    tcx,
+                    transform_predicates(tcx, predicates, options),
+                    tcx.lifetimes.re_erased,
+                    *kind,
+                );
+            }
         }
 
-        ty::Bound(..) | ty::Error(..) | ty::Infer(..) | ty::Param(..) | ty::Placeholder(..) => {
+        ty::Bound(..)
+        | ty::Error(..)
+        | ty::CoroutineClosure(..)
+        | ty::CoroutineWitness(..)
+        | ty::Infer(..)
+        | ty::Placeholder(..) => {
             bug!("transform_ty: unexpected `{:?}`", ty.kind());
         }
     }
 
     ty
 }
 
+/// Returns a type metadata identifier for the specified drop FnAbi using the Itanium C++ ABI with
+/// vendor extended type qualifiers and types for Rust types that are not used at the FFI boundary.
+pub fn typeid_for_drop_fnabi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    fn_abi: &FnAbi<'tcx, Ty<'tcx>>,
+    options: TypeIdOptions,
+) -> String {
+    // Transform self into a Drop trait object to match what the drop method's self is transformed
+    // into in transform_fnabi.
+    let self_ty = if fn_abi.args[0].layout.ty.is_mutable_ptr() {
```

Comment:
Specifically, why does `*mut _` need to be projected to `&mut dyn Drop`? Doesn't this spell for concern when we're dropping the wrong type via the wrong drop shim?

Reply:
With the information I provided above, let me know if it makes sense. It's to make both the declaration/definition label (i.e., the first call to `typeid_for_fnabi`) match the call site label (i.e., the second call to `typeid_for_fnabi`) with the information available at the latter.